### PR TITLE
Fix metrics in macro average

### LINKED ISSUE #295: Wrong numbers for multiclass acc/prec/f1 if some class is not listed in true/pred tensors
## 🐛 Bug

TM's output for precision, accuracy, f1 differs from sklearn' version of same metrics with average='macro' setting used and labels in pred/true tensors for all classes except one

### To Reproduce

1. Get one tensor with classlabels as y_pred, copy it to y_true, and pull thru torchmetrics.functional.f1/accuracy/precision with average="macro", num_classes=number of your classes used for that.
2. Convert that tensors to lists and pass into sklearn metrics, of course with average=macro where its possible
3. Compare numbers

#### Code sample

```python
y_pred = y_true = torch.tensor([1, 6, 6, 6, 3, 6, 3, 6, 6, 3, 6, 3, 6, 6, 3, 6, 3, 3, 6, 6, 6, 6, 6, 6,
        6, 3, 5, 6, 6, 3, 6, 6, 6, 6, 3, 6, 6, 3, 3, 1, 3, 1, 6, 3, 3, 1, 3, 6,
        4, 6, 6, 6, 6, 6, 6, 3, 3, 6, 3, 6, 1, 0, 5, 3, 6, 6, 6, 6, 3, 0, 6, 3,
        3, 3, 6, 3, 4]) # no class with id=2 !
print(f'Metrics:')
print(f'F1: {f1_score(y_true.numpy().astype(int).tolist(), y_pred.numpy().astype(int).tolist(), average="macro")}')
print(f'ACC: {accuracy_score(y_true.numpy().astype(int).tolist(), y_pred.numpy().astype(int).tolist())}')
print(f'PREC: {precision_score(y_true.numpy().astype(int).tolist(), y_pred.numpy().astype(int).tolist(), average="macro")}')
print(f'TM F1: {torchmetrics.functional.f1(y_pred, y_true, average="macro", num_classes=7)}')
print(f'TM acc: {torchmetrics.functional.accuracy(y_pred, y_true, average="macro", num_classes=7)}')
print(f'TM prec: {torchmetrics.functional.precision(y_pred, y_true, average="macro", num_classes=7)}')
```
```
Metrics:
F1: 1.0
ACC: 1.0
PREC: 1.0
TM F1: 0.8571428656578064
TM acc: 0.8571428656578064
TM prec: 0.8571428656578064
```

### Expected behavior

They're should be 1.0, similar to sklearn version
But adding even single item with id=2 makes numbers right, so its definitely a bug when label is exist but no examples for it in pred and true tensors

### Environment

* CUDA:
        - GPU:
                - NVIDIA Quadro RTX 5000
        - available:         True
        - version:           11.0
* Packages:
        - numpy:             1.20.3
        - pyTorch_debug:     True
        - pyTorch_version:   1.7.0+cu110
        - pytorch-lightning: 1.3.4
        - torchmetrics: 0.3.2
        - tqdm:              4.51.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - 
        - processor:         x86_64
        - python:            3.7.10
        - version:           #1 SMP Sun Feb 14 18:10:38 EST 2021

### LINKED ISSUE #300: F1 Macro score wrong(?) when label never appears
## 🐛 Bug

When F1 (Macro) is calculated on a batch where one (or more) labels never appear, the result is <1.0 even when all predictions are correct. This seems very counterintuitive. Is it expected?

### To Reproduce

```
import torch
from torchmetrics import F1

f1=F1(num_classes=2, average='macro')(preds=torch.tensor([0,0]), target=torch.tensor([0,0]))
print(f1)
```
Output: `0.5`.

This occurs if one of my batches lacks one (or more) of the labels (my real use-case has many more classes). 

### Expected behavior

I'm not an expert on how F1 is calculated, but I would expect the output to be `1.0`.

### Environment

```
pytorch-lightning==1.3.1
torch==1.8.1+cu111
torchmetrics==0.2.0
torchvision==0.9.1+cu111
```

### PR DESCRIPTION
**Pull request by vatch123**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes #295 and Fixes #300.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by vatch123** [comment]
Hey @SkafteNicki changing the weights to account for missing classes seems to fix the accuracy, and precision but not f1 score. Have a look at the following snippet.
```python
import torch
import torchmetrics
from sklearn.metrics import f1_score, accuracy_score, precision_score

y_pred = y_true = torch.tensor([1, 6, 6, 6, 3, 6, 3, 6, 6, 3, 6, 3, 6, 6, 3, 6, 3, 3, 6, 6, 6, 6, 6, 6,
        6, 3, 5, 6, 6, 3, 6, 6, 6, 6, 3, 6, 6, 3, 3, 1, 3, 1, 6, 3, 3, 1, 3, 6,
        4, 6, 6, 6, 6, 6, 6, 3, 3, 6, 3, 6, 1, 0, 5, 3, 6, 6, 6, 6, 3, 0, 6, 3,
        3, 3, 6, 3, 4]) # no class with id=2 !
print(f'Metrics:')
print(f'F1: {f1_score(y_true.numpy().astype(int).tolist(), y_pred.numpy().astype(int).tolist(), average="macro")}')
print(f'ACC: {accuracy_score(y_true.numpy().astype(int).tolist(), y_pred.numpy().astype(int).tolist())}')
print(f'PREC: {precision_score(y_true.numpy().astype(int).tolist(), y_pred.numpy().astype(int).tolist(), average="macro")}')
print(f'TM F1: {torchmetrics.functional.f1(y_pred, y_true, average="macro", num_classes=7)}')
print(f'TM acc: {torchmetrics.functional.accuracy(y_pred, y_true, average="macro", num_classes=7)}')
print(f'TM prec: {torchmetrics.functional.precision(y_pred, y_true, average="macro", num_classes=7)}')
```
```
Metrics:
F1: 1.0
ACC: 1.0
PREC: 1.0
TM F1: 0.8571428656578064
TM acc: 1.0
TM prec: 1.0
```
I would look into the f1 score issue tomorrow. I guess we should test this.

**Comment by SkafteNicki** [comment]
Hi @vatch123,
Could you try deleting this line of code:
https://github.com/PyTorchLightning/metrics/blob/1841cad3839f5d1907a1bb8bb6a266de5c5333f9/torchmetrics/functional/classification/f_beta.py#L52
I am pretty sure it is a redundant line of code which is messing things up

**Comment by vatch123** [comment]
> Hi @vatch123,
> Could you try deleting this line of code:
> https://github.com/PyTorchLightning/metrics/blob/1841cad3839f5d1907a1bb8bb6a266de5c5333f9/torchmetrics/functional/classification/f_beta.py#L52
> 
> 
> I am pretty sure it is a redundant line of code which is messing things up

Yes doing that seems to fix it.

**Comment by Borda** [comment]
@SkafteNicki @vatch123 how is it going here? :]

**Comment by vatch123** [comment]
> @SkafteNicki @vatch123 how is it going here? :]

Hey sorry was a bit busy this weekend. However, removing the line with @SkafteNicki suggested breaks the accuracy tests. So something odd is happening which I wasn't able to figure out then. Let me have a look again

**Comment by vatch123** [comment]
Hey, @SkafteNicki I was probing the issue. On local testing this particular test fails `test_accuracy.py::test_average_accuracy[preds4-target4-4-0.041666666666666664-macro-samplewise]` that is due to the following assertion error
```
===================================================== FAILURES =====================================================
__________________ test_average_accuracy[preds4-target4-4-0.041666666666666664-macro-samplewise] ___________________

preds = tensor([[[[0.1000, 0.1000, 0.1000],
          [0.2000, 0.2000, 0.2000],
          [0.3000, 0.3000, 0.3000],
          ...0.1000],
          [0.2000, 0.2000, 0.2000],
          [0.3000, 0.3000, 0.3000],
          [0.4000, 0.4000, 0.4000]]]])
target = tensor([[[1, 1, 0],
         [2, 2, 2],
         [3, 3, 3]],

        [[2, 2, 0],
         [1, 1, 1],
         [0, 0, 0]]])
num_classes = 4, exp_result = 0.041666666666666664, average = 'macro', mdmc_average = 'samplewise'

    @pytest.mark.parametrize(
        "preds, target, num_classes, exp_result, average, mdmc_average",
        [
            (_topk_preds_mcls, _topk_target_mcls, 4, 1 / 4, "macro", None),
            (_topk_preds_mcls, _topk_target_mcls, 4, 1 / 6, "weighted", None),
            (_topk_preds_mcls, _topk_target_mcls, 4, [0., 0., 0., 1.], "none", None),
            (_topk_preds_mcls, _topk_target_mcls, 4, 1 / 6, "samples", None),
            (_topk_preds_mdmc, _topk_target_mdmc, 4, 1 / 24, "macro", "samplewise"),
            (_topk_preds_mdmc, _topk_target_mdmc, 4, 1 / 6, "weighted", "samplewise"),
            (_topk_preds_mdmc, _topk_target_mdmc, 4, [0., 0., 0., 1 / 6], "none", "samplewise"),
            (_topk_preds_mdmc, _topk_target_mdmc, 4, 1 / 6, "samples", "samplewise"),
            (_topk_preds_mdmc, _topk_target_mdmc, 4, 1 / 6, "samples", "global"),
            (_av_preds_ml, _av_target_ml, 4, 5 / 8, "macro", None),
            (_av_preds_ml, _av_target_ml, 4, 0.70000005, "weighted", None),
            (_av_preds_ml, _av_target_ml, 4, [1 / 2, 1 / 2, 1., 1 / 2], "none", None),
            (_av_preds_ml, _av_target_ml, 4, 5 / 8, "samples", None),
        ],
    )
    def test_average_accuracy(preds, target, num_classes, exp_result, average, mdmc_average):
        acc = Accuracy(num_classes=num_classes, average=average, mdmc_average=mdmc_average)
    
        for batch in range(preds.shape[0]):
            acc(preds[batch], target[batch])
    
>       assert (acc.compute() == tensor(exp_result)).all()
E       assert tensor(False)
E        +  where tensor(False) = <built-in method all of Tensor object at 0x7f04240c6f40>()
E        +    where <built-in method all of Tensor object at 0x7f04240c6f40> = tensor(0.1667) == tensor(0.0417).all
E        +      where tensor(0.1667) = <function Accuracy.compute at 0x7f04b3c2e0d0>()
E        +        where <function Accuracy.compute at 0x7f04b3c2e0d0> = Accuracy().compute
E        +      and   tensor(0.0417) = tensor(0.041666666666666664)

test_accuracy.py:294: AssertionError
```
Seems like the calculated accuracy is getting rounded off somewhere. But the trace shows one more `tensor(0.1667)`. Am I interpreting it correctly? The entire test suit is taking lot of time to run at my end don't know why though.

**Comment by maximsch2** [comment]
As another comment, can you please also add the test case you used to find the issue to the tests to avoid regressions in the future?

For test suit being slow - you should be able to run just individual test files in pytest while you are debugging a particular test to avoid waiting for everything to run.

**Comment by vatch123** [comment]
Hey @SkafteNicki, I am a little stuck here. I can't figure out which inputs are breaking the tests on F1. The tests don't run locally completely throwing this error - `RuntimeError: Application timeout caused pair closure` and on the CI also I can only see the tests which break no other details as CI runs don't complete. I am assuming that the names of the tests such as `test_fbeta_f1[True-True-preds6-target6-5-None-None-_sk_fbeta_f1-None-macro-metric_class0-metric_fn0-sk_fn0]` have some information about the inputs of the test but I am unable to correlate it completely.

**Comment by Borda** [comment]
@SkafteNicki @vatch123 any chance we get for tomorrow bugfix release?

**Comment by vatch123** [comment]
Hi, Sorry for the late reply I am little stuck here and actually unable to figure what is actually breaking. I guess I would need some more inputs on this.

> Hey @SkafteNicki, I am a little stuck here. I can't figure out which inputs are breaking the tests on F1. The tests don't run locally completely throwing this error - `RuntimeError: Application timeout caused pair closure` and on the CI also I can only see the tests which break no other details as CI runs don't complete. I am assuming that the names of the tests such as `test_fbeta_f1[True-True-preds6-target6-5-None-None-_sk_fbeta_f1-None-macro-metric_class0-metric_fn0-sk_fn0]` have some information about the inputs of the test but I am unable to correlate it completely.

File: tests/classification/test_accuracy.py [hunk 1]
```diff
@@ -57,27 +58,17 @@ def _sk_accuracy(preds, target, subset_accuracy):
...
+     (_input_binary_prob.preds, _input_binary_prob.target, False), (_input_binary.preds, _input_binary.target, False),
+     (_input_mlb_prob.preds, _input_mlb_prob.target, True), (_input_mlb_logits.preds, _input_mlb_logits.target, False),
+     (_input_mlb_prob.preds, _input_mlb_prob.target, False), (_input_mlb.preds, _input_mlb.target, True),
+     (_input_mlb.preds, _input_mlb.target, False), (_input_mcls_prob.preds, _input_mcls_prob.target, False),
+     (_input_mcls_logits.preds, _input_mcls_logits.target, False), (_input_mcls.preds, _input_mcls.target, False),
+     (_input_mdmc_prob.preds, _input_mdmc_prob.target, False), (_input_mdmc_prob.preds, _input_mdmc_prob.target, True),
+     (_input_mdmc.preds, _input_mdmc.target, False), (_input_mdmc.preds, _input_mdmc.target, True),
+     (_input_mlmd_prob.preds, _input_mlmd_prob.target, True), (_input_mlmd_prob.preds, _input_mlmd_prob.target, False),
+     (_input_mlmd.preds, _input_mlmd.target, True), (_input_mlmd.preds, _input_mlmd.target, False),
+     (_input_miss_class.preds, _input_miss_class.target, False)],
```
Comment (Borda) [inline tests/classification/test_accuracy.py:71]:
```suggestion
     (_input_miss_class.preds, _input_miss_class.target, False),],
```

File: torchmetrics/functional/classification/f_beta.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ def _fbeta_compute(
 
     num = (1 + beta**2) * precision * recall
     denom = beta**2 * precision + recall
-    denom[denom == 0.] = 1  # avoid division by 0
+    # denom[denom == 0.] = 1  # avoid division by 0
```
Comment (Borda) [inline torchmetrics/functional/classification/f_beta.py:51]:
```suggestion
```

**Comment by SkafteNicki** [comment]
Hi @vatch123
Finally got around to debug this and it took longer time than expected. It was essentially the fbeta metric that was making trouble, because it is a bit different from the other metrics due to its nature of not being directly computed from `tp, fp, tn, tp` but instead from `precision` and `recall` and therefore had to be dealt with in a special way. Hopefully all test should pass now :]

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
> Finally got around to debug this and it took longer time than expected.

seems that some tests are still failing :]

**Comment by vatch123** [comment]
> > Finally got around to debug this and it took longer time than expected.
> 
> seems that some tests are still failing :]

Yes, some `precision-recall` and `same input` tests are breaking. 
On a side note, can someone explain how the tests are named? What is the convention? Do these name carry any meaning `test_precision_recall_fn[preds6-target6-5-None-None-_sk_prec_recall-0-macro-Precision-precision-precision_score]`, `test_fbeta_f1[True-True-preds6-target6-5-None-None-_sk_fbeta_f1-None-macro-metric_class0-metric_fn0-sk_fn0]`. I am assuming they do.

**Comment by SkafteNicki** [comment]
Okay, after a small update, everything passes locally for me now :]

@vatch123 about naming lets break `test_precision_recall_fn[preds6-target6-5-None-None-_sk_prec_recall-0-macro-Precision-precision-precision_score]` down
1. `test_precision_recall_fn` is the function being called, so https://github.com/PyTorchLightning/metrics/blob/0495776dc67ebc761fe35bf8c6311d8987523257/tests/classification/test_precision_recall.py#L259 in this case. The function name should be read as "test the functional (fn) version of precision recall"
2. everything inside the square brackets is what parametrize the test, in this case https://github.com/PyTorchLightning/metrics/blob/0495776dc67ebc761fe35bf8c6311d8987523257/tests/classification/test_precision_recall.py#L173-L202. Note that the parametrizations are listed in opposite order so `preds6-target6-5-None-None-_sk_prec_recall` corresponds to the last parametrization call.

**Comment by vatch123** [comment]
> Okay, after a small update, everything passes locally for me now :]
> 
> @vatch123 about naming lets break `test_precision_recall_fn[preds6-target6-5-None-None-_sk_prec_recall-0-macro-Precision-precision-precision_score]` down
> 
> 1. `test_precision_recall_fn` is the function being called, so https://github.com/PyTorchLightning/metrics/blob/0495776dc67ebc761fe35bf8c6311d8987523257/tests/classification/test_precision_recall.py#L259
>     in this case. The function name should be read as "test the functional (fn) version of precision recall"
> 2. everything inside the square brackets is what parametrize the test, in this case https://github.com/PyTorchLightning/metrics/blob/0495776dc67ebc761fe35bf8c6311d8987523257/tests/classification/test_precision_recall.py#L173-L202
>    . Note that the parametrizations are listed in opposite order so `preds6-target6-5-None-None-_sk_prec_recall` corresponds to the last parametrization call.

Thanks for this.
Yes everything runs well locally at my end too. But some tests are skipped locally maybe one of those is breaking.

**Comment by Borda** [comment]
> Yes everything runs well locally at my end too. But some tests are skipped locally maybe some of those are breaking.

yes, you can simply check the skipped test if you run with `-v` and compare what is skipped on CI and locally

**Comment by vatch123** [comment]
Something odd is happening. This test `test_precision_recall_fn[preds6-target6-5-None-None-_sk_prec_recall-0-weighted-Precision-precision-precision_score]` fails on the CI.
image

but the same test
`test_precision_recall_fn[preds6-target6-5-None-None-_sk_prec_recall-0-weighted-Precision-precision-precision_score]` passes locally
image
Can there be an issue on the CI or on my local deployment?

**Comment by Borda** [comment]
> Can there be an issue on the CI or on my local deployment?

not sure, what is your local environment - OS, python, packages?

you can use docker image to be sure you use the same, see
https://github.com/PyTorchLightning/metrics/blob/5c2069eb2087ede7f008222ebab286aedad14dfb/azure-pipelines.yml#L29

and proceed following steps:
https://github.com/PyTorchLightning/metrics/blob/5c2069eb2087ede7f008222ebab286aedad14dfb/azure-pipelines.yml#L55-L59
and test as 
https://github.com/PyTorchLightning/metrics/blob/5c2069eb2087ede7f008222ebab286aedad14dfb/azure-pipelines.yml#L68

**Comment by vatch123** [comment]
> > Can there be an issue on the CI or on my local deployment?
> 
> not sure, what is your local environment - OS, python, packages?
> 
I am using WSL2 with ubuntu 20.04 64-bit, python 3.8.8, torch 1.8

**Comment by Borda** [comment]
> I am using WSL2 with ubuntu 20.04 64-bit, python 3.8.8, torch 1.8

can you run it in docker as https://github.com/PyTorchLightning/metrics/pull/303#issuecomment-882019475

**Comment by vatch123** [comment]
> > I am using WSL2 with ubuntu 20.04 64-bit, python 3.8.8, torch 1.8
> 
> can you run it in docker as [#303 (comment)](https://github.com/PyTorchLightning/metrics/pull/303#issuecomment-882019475)

Yes, sure. I tried yesterday. I got the docker image downloaded and started a bash process in it to run the above commands. But, I am not very well versed with docker so I could not get it to run. The image threw an error that python itself was not found. I must have made some error somewhere. I will look into it again. If you have some further insights that might be helpful it would be great. Thanks

**Comment by vatch123** [comment]
Okay something weird happened. When I run the entire test suite using `python -m pytest torchmetrics tests -v --cov=torchmetrics --junitxml=$(Build.StagingDirectory)/test-results.xml --durations=50` some tests fail. One example of that (I could not run the entire test suite as it was taking a lot of time)
image
As we can a precision-recall test is failing when `macro` is passed.

But, now if run the only the precision-recall file using `python -m pytest torchmetrics tests/classification/test_precision_recall.py -v --cov=torchmetrics --junitxml=$(Build.StagingDirectory)/test-results.xml --durations=50`
image
The same test passes. In fact all the tests pass when I run a single file.

Can there be a reason for this? Maybe some configurations are set while running the entire suite?

**Comment by Borda** [comment]
the before and after are perform equally for running all and just some files testing...
so it may be some import correlation, or some other actor changes its status :/
you can enter the test with a debugger and check what is the difference between these two calls :rabbit:

**Comment by Borda** [comment]
@vatch123 how is it going here? :rabbit:

**Review by antonibigata — APPROVED** [review]

**Comment by vatch123** [comment]
> @vatch123 how is it going here? 🐰

Hey. Sorry been busy recently did not get a chance to probe deeper. I tried with a debugger but haven't been able to pinpoint the error. Will need some time and help on this I guess.

**Review by SkafteNicki — APPROVED** [review]

**Comment by vatch123** [comment]
Hey @SkafteNicki. Thanks for the help here. Can you let me know what was the actual issue? I see you mentioned `dist_sync_on_step` not working and you added one more test. Can you just briefly describe the issue? Thanks again.

**Pull request merged**